### PR TITLE
Decrease the  num_stars when deleting a repo

### PR DESCRIPTION
**Pull request by a1012112796**
As title.

### CODE REVIEW

**Comment by lafriks** [comment]
I don't think this fixes described issue

**Comment by a1012112796** [comment]
> I don't think this fixes described issue
Reason? Thanks

**Comment by a1012112796** [comment]
@lafriks Maybe should recalculate  the stars number when deieteing repo. is it right ? Thanks

**Comment by a1012112796** [comment]
Hmm, Maybe should add migration to recalculate all , recalculate stars  when deleting repo will cost too much time ..

**Comment by lafriks** [comment]
Migration should be batched, see other migrations for examples

File: models/migrations/v143.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/log"
+	"xorm.io/xorm"
```
Comment (lafriks) [inline models/migrations/v143.go:10]:
```suggestion

	"xorm.io/xorm"
```

File: models/migrations/v143.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+		var number int64
+
+		for _, uid := range userIDs {
+			if number, err = x.Where("uid = ?", uid).Count(new(models.Star)); err != nil {
+				return
+			}
+
+			if _, err = x.Exec("UPDATE `user` SET num_stars=? WHERE id = ?", number, uid); err != nil {
+				return err
+			}
```
Comment (lafriks) [inline models/migrations/v143.go]:
```suggestion
			if _, err = x.Exec("UPDATE `user` SET num_stars=(SELECT COUNT(*) FROM `star` WHERE uid=?) WHERE id=?", uid, uid); err != nil {
				return err
			}
			
```

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
We should backport this without migration, only bugfix itself

**Review by CirnoT — APPROVED** [review]

**Comment by CirnoT** [comment]
Actually, is `num_stars` even used anywhere? I see that we have `GetStarredRepoCount` which ignores it completely.

**Comment by a1012112796** [comment]
> Actually, is `num_stars` even used anywhere? I see that we have `GetStarredRepoCount` which ignores it completely.

@CirnoT Yes, because count repo numbers will use too much time if users have many rpos. So it's necessary to save it to db. And the `GetStarredRepoCount` is not used now. 

image

**Comment by lunny** [comment]
Maybe become as part of doctor is better and easy to back port.

**Comment by a1012112796** [comment]
> Maybe become as part of doctor is better and easy to back port.

Do you means follow #9095 ? Thounds great, I will try. Thanks

**Comment by a1012112796** [comment]
https://github.com/go-gitea/gitea/blob/5b1327eb56b29218f6c37c22042603a7824a7a9f/cmd/doctor.go#L76-L125

Maybe should split them into more files like migration, or this file will become too big ....

**Comment by a1012112796** [comment]
@lunny  How about this change? Thanks

File: models/migrations/v143.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+	"code.gitea.io/gitea/models"
+
+	"xorm.io/xorm"
+)
+
+func recalculateStars(x *xorm.Engine) (err error) {
+	// because of issue https://github.com/go-gitea/gitea/issues/11949,
+	// recalculate Stars number for all users to fully fix it.
+
+	return models.DoctorUserStarNum()
```
Comment (lafriks) [inline models/migrations/v143.go]: It's better to duplicate code here as there is no guarantee that model won't change in the future and that could break migration

Comment (jolheiser) [reply models/migrations/v143.go]: `gitea-vet` will also fail CI if it detects models package in a migration in the future.  
https://gitea.com/gitea/gitea-vet/pulls/5 (not released yet)

File: models/migrations/v143.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+		users := make([]User, 0, batchSize)
+		if err = sess.Limit(batchSize, start).Where("type = ?", 0).Cols("id").Find(&users); err != nil {
+			return err
+		}
+		if len(users) == 0 {
+			break
+		}
+
+		for _, user := range users {
+			if _, err = x.Exec("UPDATE `user` SET num_stars=(SELECT COUNT(*) FROM `star` WHERE uid=?) WHERE id=?", user.ID, user.ID); err != nil {
```
Comment (lunny) [inline models/migrations/v143.go]: sess.Exec

**Review by lunny — CHANGES_REQUESTED** [review]
Also for a big instance which has many users, the transaction maybe failed. So it's more reasonable to commit every 100 or 200 updates.

File: models/migrations/v143.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+	for start := 0; ; start += batchSize {
+		users := make([]User, 0, batchSize)
+		if err = sess.Limit(batchSize, start).Where("type = ?", 0).Cols("id").Find(&users); err != nil {
+			return
+		}
+		if len(users) == 0 {
+			break
+		}
+
```
Comment (lunny) [inline models/migrations/v143.go:33]: Sorry, you also need a `sess.Begin()` to start a transaction.

File: models/repo.go [hunk 6]
```diff
@@ -2351,3 +2355,34 @@ func updateRepositoryCols(e Engine, repo *Repository, cols ...string) error {
...
+	for start := 0; ; start += batchSize {
+		users := make([]User, 0, batchSize)
+		if err = sess.Limit(batchSize, start).Where("type = ?", 0).Cols("id").Find(&users); err != nil {
+			return
+		}
+		if len(users) == 0 {
+			break
+		}
+
+		for _, user := range users {
```
Comment (lunny) [inline models/repo.go:2368]: As above.

File: models/migrations/v143.go (hunk 5 above)
Comment (a1012112796) [reply models/migrations/v143.go:33]: Sorry, I forgot it.

**Review by lafriks — APPROVED** [review]

**Comment by a1012112796** [comment]
@lunny Please review again, Thanks

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
make L-G-T-M work.

**Comment by techknowlogick** [comment]
@a1012112796 thanks for PR :) Please send backport to release/v1.12 branch

**Pull request merged**